sdk/framework/openapi.go

```diff
@@ -521,18 +509,65 @@ func documentPath(p *Path, backend *Backend, requestResponsePrefix string, doc *
 			switch opType {
 			case logical.CreateOperation, logical.UpdateOperation:
 				pi.Post = op
-			case logical.ReadOperation, logical.ListOperation:
+			case logical.ReadOperation:
 				pi.Get = op
 			case logical.DeleteOperation:
 				pi.Delete = op
+			case logical.ListOperation:
+				listOperation = op
 			}
 		}
 
-		openAPIPath := "/" + path
-		if doc.Paths[openAPIPath] != nil {
-			backend.Logger().Warn("OpenAPI spec generation: multiple framework.Path instances generated the same path; last processed wins", "path", openAPIPath)
+		// Write the regular, non-list, OpenAPI path to the OpenAPI document, UNLESS we generated a ListOperation, and
+		// NO OTHER operation types. In that fairly common case (there are lots of list-only endpoints), we avoid
+		// writing a redundant OpenAPI path for (e.g.) "auth/token/accessors" with no operations, only to then write
+		// one for "auth/token/accessors/" immediately below.
+		//
+		// On the other hand, we do still write the OpenAPI path here if we generated ZERO operation types - this serves
+		// to provide documentation to a human that an endpoint exists, even if it has no invokable OpenAPI operations.
+		// Examples of this include kv-v2's ".*" endpoint (regex cannot be translated to OpenAPI parameters), and the
+		// auth/oci/login endpoint (implements ResolveRoleOperation only, only callable from inside Vault).
+		generateNonListOpenAPIPath := listOperation == nil || pi.Get != nil || pi.Post != nil || pi.Delete != nil
+		if generateNonListOpenAPIPath {
+			openAPIPath := "/" + path
+			if doc.Paths[openAPIPath] != nil {
+				backend.Logger().Warn(
+					"OpenAPI spec generation: multiple framework.Path instances generated the same path; "+
+						"last processed wins", "path", openAPIPath)
+			}
+			doc.Paths[openAPIPath] = &pi
+		}
+
+		// If there is a ListOperation, write it to a separate OpenAPI path in the document.
+		if listOperation != nil {
+			// Typically, we will append a slash here to disambiguate from the path written immediately above.
+			// However, if we skipped writing the path above (and so don't need to worry about disambiguating from it),
+			// AND the path already contains a trailing slash, we want to avoid doubling it.
```

Comment:
In brief:


It would be possible to make up some `framework.Path` input that would result in this code generating a "//" suffix ... but only if you define a regex pattern which ends with an unconditional slash, combined with both list and non-list operations. Such a configuration would be intrinsically broken, since a pattern ending with an unconditional slash, could never match non-list operations - therefore I'm not that worried about handling it.

Although, having written this out, I guess we could just handle this by refusing to render the impossible-to-invoke non-list operations to the OpenAPI document, and avoid the "//" in this edge case that way... I'll see if I can code that tomorrow.


More detail:

The regex -> OpenAPI translation will only include a trailing slash if there is an **unconditional** slash matching at the end of the path. Any unconditional slash at the end of the path is usually bad news, because it means the pattern will only ever match ListOperations, meaning HelpOperations will fail to match (the thing I fixed in #18571). In writing this PR, I initially included a warning log message if any generated OpenAPI patterns were found to have a trailing slash, expecting that I'd fixed them all in that PR... but then my warning log message triggered...

1) Enterprise-only (now stubbed in OSS) path `sys/storage/raft/snapshot-auto/config/` has an unconditional slash at the end:

https://github.com/hashicorp/vault/blob/d1c943b25559c98998d0985e61739c5b21283c1f/vault/logical_system_helpers.go#L224

This is a repeat of the problem I fixed for OSS paths in #18571 and presumably means that `vault path-help sys/storage/raft/snapshot-auto/config` won't work either.

Ideally this needs to be fixed in the Enterprise repo and OSS stub too, to comply with:
https://github.com/hashicorp/vault/blob/d1c943b25559c98998d0985e61739c5b21283c1f/sdk/framework/path.go#L63-L69

2) Rather more complicated, is `sys/leases/lookup`... it turns out that `sys/leases/lookup` is one endpoint, and `sys/leases/lookup/` is a completely separate endpoint originating from a different `framework.Path` struct. 

I can and should refactor this, since it's responsible for the illogical OpenAPI client generation of separate functions `LeasesLookUp` and `LeasesLookUpWithPrefix`, when any human writing an API wrapper for this would instead represent this as a single `LeasesLookUp` function, taking a path prefix, which could be a zero-length string - but I didn't want to further complicate this PR.

Reply:
> I can and should refactor this, since it's responsible for the illogical OpenAPI client generation of separate functions `LeasesLookUp` and `LeasesLookUpWithPrefix`, when any human writing an API wrapper for this would instead represent this as a single `LeasesLookUp` function, taking a path prefix, which could be a zero-length string - but I didn't want to further complicate this PR.

Aaargh. Except, this runs into the OpenAPI **spec** lacking support for truly describing this kind of API! (Currently, path parameters are assumed to be required to be present, with a non-empty value, and to not contain slashes... this last part is going to need more thought, so it's just as well to leave it out of scope for this particular PR.)

Reply:
> It would be possible to make up some `framework.Path` input that would result in this code generating a "//" suffix ... but only if you define a regex pattern which ends with an unconditional slash, combined with both list and non-list operations. Such a configuration would be intrinsically broken, since a pattern ending with an unconditional slash, could never match non-list operations - therefore I'm not that worried about handling it.
> 
> Although, having written this out, I guess we could just handle this by refusing to render the impossible-to-invoke non-list operations to the OpenAPI document, and avoid the "//" in this edge case that way... I'll see if I can code that tomorrow.

Done.

Reply:
Thanks for the detailed explanation, this is way more complicated than I imagined :)